### PR TITLE
Implement a variant of timeseries split.

### PR DESCRIPTION
**Pull request by Sahil333**
- Implements CTimeSeriesSplitting which splits time series, respecting time, for cross-validation. #3847 

- CTimeSeriesSplitting splits time-series indices at c[n/k], c = 1,2,3,...,K-1. For last split, split index is n-min_subset_size( default 1 ).

### CODE REVIEW

File: examples/undocumented/libshogun/splitting_timeseries_crossvalidation.cpp [hunk 1]
```diff
@@ -0,0 +1,100 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
```
Comment (vigsterkr) [inline examples/undocumented/libshogun/splitting_timeseries_crossvalidation.cpp]: plz use the BSD license header from here (currently we are in transition to BSD based license): 
https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/lib/any.h#L2

File: src/shogun/evaluation/TimeSeriesSplitting.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
+#include <shogun/evaluation/TimeSeriesSplitting.h>
```
Comment (vigsterkr) [inline src/shogun/evaluation/TimeSeriesSplitting.cpp]: plz add bsd license

File: src/shogun/evaluation/TimeSeriesSplitting.cpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+	for (index_t i = 0; i < num_subsets; i++)
+	{
+		CDynamicArray<index_t>* current =
+		    (CDynamicArray<index_t>*)m_subset_indices->get_element(i);
+
+		/* filling current with indices on right end  */
+		for (index_t k = i == num_subsets - 1
+		                     ? indices.vlen - m_h
+		                     : (i + 1) * (indices.vlen / num_subsets);
+		     k < indices.vlen; k++)
```
Comment (vigsterkr) [inline src/shogun/evaluation/TimeSeriesSplitting.cpp]: you could use the prefix operator here... ++k

File: src/shogun/evaluation/TimeSeriesSplitting.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+void CTimeSeriesSplitting::build_subsets()
+{
+	reset_subsets();
+	m_is_filled = true;
+
+	SGVector<index_t> indices(m_labels->get_num_labels());
+	indices.range_fill();
+	index_t num_subsets = m_subset_indices->get_num_elements();
+
+	for (index_t i = 0; i < num_subsets; i++)
```
Comment (vigsterkr) [inline src/shogun/evaluation/TimeSeriesSplitting.cpp]: prefix operator `++i`

Comment (vigsterkr) [inline src/shogun/evaluation/TimeSeriesSplitting.cpp]: and of course `auto i = 0` should be good as well

File: src/shogun/evaluation/TimeSeriesSplitting.h [hunk 5]
```diff
@@ -0,0 +1,62 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
```
Comment (vigsterkr) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: use the bsd license plz

**Comment by karlnapf** [comment]
@olinguyen this will be useful for you once merged

**Comment by karlnapf** [comment]
Will review soon

File: examples/undocumented/libshogun/splitting_timeseries_crossvalidation.cpp [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+using namespace shogun;
+
+void print_message(FILE* target, const char* str)
+{
+	fprintf(target, "%s", str);
+}
+
+int main(int argc, char** argv)
+{
+	init_shogun(&print_message, &print_message, &print_message);
```
Comment (karlnapf) [inline examples/undocumented/libshogun/splitting_timeseries_crossvalidation.cpp]: We don't add these type of examples anymore, if it is possible to express the functionality in the [meta examples](http://shogun.ml/examples)
See [here](https://github.com/shogun-toolbox/shogun/blob/develop/doc/readme/EXAMPLES.md) how to write one

File: src/shogun/evaluation/TimeSeriesSplitting.h [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+		/** constructor
+		 *
+		 * @param labels labels to be (possibly) used for splitting
+		 * @param num_subsets desired number of subsets, the labels are split
+		 * into
+		 */
+		CTimeSeriesSplitting(CLabels* labels, index_t num_subsets);
+
+		/** @param h future needed in test set. */
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: I don't understand this doc, can you make it a bit more explicit?

File: src/shogun/evaluation/TimeSeriesSplitting.h [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+#include <shogun/evaluation/SplittingStrategy.h>
+#include <shogun/lib/config.h>
+
+namespace shogun
+{
+	class CLabels;
+
+	/** @brief Implementation of timeseries splitting for cross-validation.
+	 * Produces subsets each filled with indices greater than a split point.
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: Can you mention the types of parameters that have to be set here, and maybe give an example of a splitting?

File: src/shogun/evaluation/TimeSeriesSplitting.h [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+		/** @return h value. */
+		index_t get_h();
+
+		/** @return name of the SGSerializable */
+		virtual const char* get_name() const
+		{
+			return "TimeSeriesSplitting";
+		}
+
+		/** implementation of the time-series cross-validation splitting
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: Please start docs with capital letters (they are sentences)
even if this is not the case in other parts of shogun

File: src/shogun/evaluation/TimeSeriesSplitting.h [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+		 * strategy */
+		virtual void build_subsets();
+
+		/** custom rng if using cross validation across different threads */
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: isn't this inherited from the base class? If not, it should be I think
@vigsterkr

File: src/shogun/evaluation/TimeSeriesSplitting.h [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+		CRandom* m_rng;
+
+		/** atleast h size of each subsets.(default 1) */
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: Please use proper punctuation and proper English

File: src/shogun/evaluation/TimeSeriesSplitting.h [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+		index_t m_h = 1;
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: please dont use default initialization but do that in an `init` method
Also, `h` is not a good member name. Name it so that it is clear what it does

File: src/shogun/evaluation/TimeSeriesSplitting.cpp [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+ */
+
+#include <shogun/evaluation/TimeSeriesSplitting.h>
+#include <shogun/labels/Labels.h>
+
+using namespace shogun;
+
+CTimeSeriesSplitting::CTimeSeriesSplitting() : CSplittingStrategy()
+{
+	m_rng = sg_rand;
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.cpp]: This should be done in the base class

File: src/shogun/evaluation/TimeSeriesSplitting.cpp [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+CTimeSeriesSplitting::CTimeSeriesSplitting(CLabels* labels, index_t num_subsets)
+    : CSplittingStrategy(labels, num_subsets)
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.cpp]: can you introduce a method `init` and default initialise members in there? Thats shogun style.
If you do that, you cannot use initialization fields anymore but have to overwrite num_subsets and labels after calling init

File: src/shogun/evaluation/TimeSeriesSplitting.cpp [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+	indices.range_fill();
+	index_t num_subsets = m_subset_indices->get_num_elements();
+
+	for (auto i = 0; i < num_subsets; ++i)
+	{
+		CDynamicArray<index_t>* current =
+		    (CDynamicArray<index_t>*)m_subset_indices->get_element(i);
+
+		/* filling current with indices on right end  */
+		for (auto k = i == num_subsets - 1
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.cpp]: this for loop needs some serious demangling

File: src/shogun/evaluation/TimeSeriesSplitting.cpp [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+void CTimeSeriesSplitting::set_h(index_t h)
+{
+	index_t num_subsets = m_subset_indices->get_num_elements();
+	index_t num_labels = m_labels->get_num_labels();
+
+	/* h value should not be greater than difference between number of labels
+	 * and start index of second last split point */
+	if (h >= num_labels - (num_subsets - 1) * (num_labels / num_subsets))
+	{
+		SG_WARNING("h can not be %d. Setting h value to default 1.\n", h);
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.cpp]: The message should be "Number XXX (%d) has to be at least %d"

Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.cpp]: make this a REQUIRE. giving warnings and then taking implicit decisions (h=1) is harmful

File: tests/unit/evaluation/SplittingStrategy_unittest.cc [hunk 17]
```diff
@@ -296,3 +296,69 @@ TEST(SplittingStrategy,LOO)
 		SG_UNREF(splitting);
 	}
 }
+
+TEST(SplittingStrategy, timeseries_split_linear)
```
Comment (karlnapf) [inline tests/unit/evaluation/SplittingStrategy_unittest.cc]: this unit test is way too long
Can you split it into multiple ones?

Comment (karlnapf) [inline tests/unit/evaluation/SplittingStrategy_unittest.cc]: Also each of the unit tests should have a name that explains what is tested

File: tests/unit/evaluation/SplittingStrategy_unittest.cc [hunk 18]
```diff
@@ -296,3 +296,69 @@ TEST(SplittingStrategy,LOO)
...
+{
+	index_t num_labels, num_subsets, h_value;
+	index_t runs = 10;
+
+	while (runs-- > 0)
+	{
+		num_labels = CMath::random(50, 150);
+		num_subsets = CMath::random(1, 10);
+		h_value = CMath::random(1, 20);
+		/* build labels */
```
Comment (karlnapf) [inline tests/unit/evaluation/SplittingStrategy_unittest.cc]: superfluous comment

File: tests/unit/evaluation/SplittingStrategy_unittest.cc [hunk 19]
```diff
@@ -296,3 +296,69 @@ TEST(SplittingStrategy,LOO)
...
+		CRegressionLabels* labels = new CRegressionLabels(num_labels);
+		for (index_t i = 0; i < num_labels; ++i)
+			labels->set_label(i, CMath::random(-10.0, 10.0));
+		/* build linearly increasing split strategy */
+		CTimeSeriesSplitting* splitting =
+		    new CTimeSeriesSplitting(labels, num_subsets);
+		/* set future h needed in test */
+		splitting->set_h(h_value);
+		/* build subset */
```
Comment (karlnapf) [inline tests/unit/evaluation/SplittingStrategy_unittest.cc]: all of those comments are superflous

File: src/shogun/evaluation/TimeSeriesSplitting.h (hunk 10 above)
Comment (Sahil333) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: yes, it should be inherited but, without it, the compiler gives error "build_subsets() is not defined". So, I left it there and compiled successfully.

File: src/shogun/evaluation/TimeSeriesSplitting.h (hunk 9 above)
Comment (Sahil333) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: Thanks, I'll take care of that.

File: src/shogun/evaluation/TimeSeriesSplitting.h (hunk 12 above)
Comment (Sahil333) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: I'll change it to min_subset_size.

File: src/shogun/evaluation/TimeSeriesSplitting.cpp (hunk 13 above)
Comment (Sahil333) [inline src/shogun/evaluation/TimeSeriesSplitting.cpp]: I'll initiate m_rng and m_min_subset_size in init()?

File: src/shogun/evaluation/TimeSeriesSplitting.h (hunk 8 above)
Comment (Sahil333) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: Peace, I'll mention it more precisely.

File: examples/undocumented/libshogun/splitting_timeseries_crossvalidation.cpp (hunk 6 above)
Comment (Sahil333) [inline examples/undocumented/libshogun/splitting_timeseries_crossvalidation.cpp]: Okay, Let me see.

File: src/shogun/evaluation/TimeSeriesSplitting.h (hunk 7 above)
Comment (Sahil333) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: Right, I'll make it more explicit

File: examples/undocumented/libshogun/splitting_timeseries_crossvalidation.cpp (hunk 6 above)
Comment (Sahil333) [inline examples/undocumented/libshogun/splitting_timeseries_crossvalidation.cpp]: Can you tell me where can I ask my specific questions about shogun? When I am stuck somewhere in shogun library.

Comment (vigsterkr) [inline examples/undocumented/libshogun/splitting_timeseries_crossvalidation.cpp]: @Sahil333 there's the IRC group or of course you can ask on the mailing list

File: src/shogun/evaluation/TimeSeriesSplitting.h (hunk 10 above)
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: "build_subsets()" is not defined? How is that related to the rng?
Anyways, would you mind moving this to the base class and inheriting from there?

File: src/shogun/evaluation/TimeSeriesSplitting.h (hunk 12 above)
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: Any updates? :)

File: src/shogun/evaluation/TimeSeriesSplitting.cpp (hunk 13 above)
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.cpp]: yes

**Comment by Sahil333** [comment]
Python meta examples can't import WrappedObjectArray. I tried the build from my feature branch as well as develop branch. I don't know why is it happening.
 
test 691
    Start 691: generated_python-evaluation-cross_validation_mkl_weight_storage

691: Test command: /usr/bin/python "/media/sahil333/MostlyHere/GSOC/shogun/build/examples/meta/python/evaluation/cross_validation_mkl_weight_storage.py"
691: Environment variables: 
691:  GDB="gdb -x /media/sahil333/MostlyHere/GSOC/shogun/src/.gdb --args"
691:  PYTHON=/usr/bin/python
691:  PYTHONPATH=/usr/local/lib/python2.7/dist-packages:
691:  LD_LIBRARY_PATH=/usr/local/lib:
691: Test timeout computed to be: 1500
691: Traceback (most recent call last):
691:   File "/media/sahil333/MostlyHere/GSOC/shogun/build/examples/meta/python/evaluation/cross_validation_mkl_weight_storage.py", line 16, in <module>
691:     from shogun import WrappedObjectArray
691: ImportError: cannot import name WrappedObjectArray

**Comment by karlnapf** [comment]
Sorry for the delay! I think this was fixed now. If you rebase, what happens?
Shall we aim to get this in soon?

**Comment by Sahil333** [comment]
My hard disk was broken so had to rebuild. API examples and integration tests are passing for C++. But when I am testing API examples for python interface it is giving following error now. Should I use "make install" to install shogun.py systemwide? Or how can I set PYTHONPATH to /build/src/interface/python/shogun.py?

test 756
      Start 756: generated_python-gaussian_processes-sparse_gaussian_process_regression

756: Test command: /usr/bin/python "/home/happy/shogun/build/examples/meta/python/gaussian_processes/sparse_gaussian_process_regression.py"
756: Environment variables: 
756:  GDB="gdb -x /home/happy/shogun/src/.gdb --args"
756:  PYTHON=/usr/bin/python
756:  PYTHONPATH=/usr/local/lib/python2.7/dist-packages:
756:  LD_LIBRARY_PATH=/usr/local/lib:
756: Test timeout computed to be: 9.99988e+06
756: Traceback (most recent call last):
756:   File "/home/happy/shogun/build/examples/meta/python/gaussian_processes/sparse_gaussian_process_regression.py", line 2, in <module>
756:     from shogun import CSVFile
756: ImportError: No module named shogun
53/53 Test #756: generated_python-gaussian_processes-sparse_gaussian_process_regression ...***Failed    0.12 sec

**Comment by karlnapf** [comment]
Check http://shogun.ml/examples/latest/quickstart/interfaces.html

**Comment by karlnapf** [comment]
Actually, @lisitsyn did changed to testsuite to use the build output in #4034 
So now it should work if you rebase against the latest develop

Let's get this one done, it is very cool to have!

File: src/shogun/evaluation/TimeSeriesSplitting.h [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+	 * The two splits are \f$ S1 = [5,6,7,8,9] \f$ and \f$ S2 = [8,9]\f$
+	 */
+
+	class CTimeSeriesSplitting : public CSplittingStrategy
+	{
+	public:
+		/** constructor */
+		CTimeSeriesSplitting();
+
+		/** constructor
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: Minor: Capital "C"

File: src/shogun/evaluation/TimeSeriesSplitting.h [hunk 21]
```diff
@@ -0,0 +1,98 @@
...
+		 *
+		 * @param labels labels to be (possibly) used for splitting
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: I think this class should not accept labels ... labels are not needed to build the splitting

File: src/shogun/evaluation/TimeSeriesSplitting.h [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+		 * @param num_subsets desired number of subsets, the labels are split
+		 * into
+		 */
+		CTimeSeriesSplitting(CLabels* labels, index_t num_subsets);
+
+		/** Sets the minimum subset size for subsets. If forecasting h-step
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: Could we talk about "folds" and "fold size" here, rather than subsets?

Also, I think this method should be called different "num_steps_future" for example

File: src/shogun/evaluation/TimeSeriesSplitting.h [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+		/** @return name of the SGSerializable */
+		virtual const char* get_name() const
+		{
+			return "TimeSeriesSplitting";
+		}
+
+		void build_subsets() override;
+
+		/** Custom rng if using cross validation across different threads */
+		CRandom* m_rng;
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: Can we please move this one into the base class?

File: tests/unit/evaluation/SplittingStrategy_unittest.cc [hunk 24]
```diff
@@ -296,3 +296,90 @@ TEST(SplittingStrategy,LOO)
...
 }
+
+TEST(SplittingStrategy, timeseries_subset_linear_splits)
+{
+	index_t num_labels, num_subsets, min_subset_size, base_size;
+	index_t runs = 10;
+
+	while (runs-- > 0)
+	{
+		num_labels = CMath::random(50, 150);
```
Comment (karlnapf) [inline tests/unit/evaluation/SplittingStrategy_unittest.cc]: Could you create a mock class here, or at least a function that generates all the code that is shared across the different tests? See e.g. `tests/unit/FisherLDA_unittest.cc` on how to do that

File: tests/unit/evaluation/SplittingStrategy_unittest.cc [hunk 25]
```diff
@@ -296,3 +296,90 @@ TEST(SplittingStrategy,LOO)
 		SG_UNREF(splitting);
 	}
 }
+
+TEST(SplittingStrategy, timeseries_subset_linear_splits)
+{
+	index_t num_labels, num_subsets, min_subset_size, base_size;
+	index_t runs = 10;
+
+	while (runs-- > 0)
```
Comment (karlnapf) [inline tests/unit/evaluation/SplittingStrategy_unittest.cc]: please use a for loop
`for (auto i=0; i<runs; i++)`
or even better
`(auto i : Range(3, 10))`, see `base/range.h`

File: tests/unit/evaluation/SplittingStrategy_unittest.cc [hunk 26]
```diff
@@ -296,3 +296,90 @@ TEST(SplittingStrategy,LOO)
...
+
+		splitting->set_min_subset_size(min_subset_size);
+		splitting->build_subsets();
+
+		for (index_t i = 0; i < num_subsets; ++i)
+		{
+			SGVector<index_t> subset = splitting->generate_subset_indices(i);
+			SGVector<index_t> inverse = splitting->generate_subset_inverse(i);
+
+			/* Subset size should be atleat min_subset_size */
```
Comment (karlnapf) [inline tests/unit/evaluation/SplittingStrategy_unittest.cc]: minor "at least"

Comment (karlnapf) [inline tests/unit/evaluation/SplittingStrategy_unittest.cc]: superflous comment, clear from the code, remove

File: tests/unit/evaluation/SplittingStrategy_unittest.cc [hunk 27]
```diff
@@ -296,3 +296,90 @@ TEST(SplittingStrategy,LOO)
 		SG_UNREF(splitting);
 	}
 }
+
```
Comment (karlnapf) [inline tests/unit/evaluation/SplittingStrategy_unittest.cc]: I am missing a test where you do things/checks by hand

 * manually specify some simple data
 * manually explicitly check the folds (since you know the data)
 * really small example, yet very useful

File: tests/unit/evaluation/SplittingStrategy_unittest.cc [hunk 28]
```diff
@@ -296,3 +296,90 @@ TEST(SplittingStrategy,LOO)
...
+	index_t num_labels, num_subsets, min_subset_size;
+	index_t runs = 10;
+
+	while (runs-- > 0)
+	{
+		num_labels = CMath::random(50, 150);
+		num_subsets = CMath::random(1, 5);
+		min_subset_size = CMath::random(1, 7);
+
+		CRegressionLabels* labels = new CRegressionLabels(num_labels);
```
Comment (karlnapf) [inline tests/unit/evaluation/SplittingStrategy_unittest.cc]: again, all this should be moved to some helper function. the test is bloated and hard to read right now

File: tests/unit/evaluation/SplittingStrategy_unittest.cc [hunk 29]
```diff
@@ -296,3 +296,90 @@ TEST(SplittingStrategy,LOO)
...
+
+		splitting->set_min_subset_size(min_subset_size);
+		splitting->build_subsets();
+
+		for (index_t i = 0; i < num_subsets; ++i)
+		{
+			SGVector<index_t> subset = splitting->generate_subset_indices(i);
+			SGVector<index_t> inverse = splitting->generate_subset_inverse(i);
+
+			/* check future leak into test set */
```
Comment (karlnapf) [inline tests/unit/evaluation/SplittingStrategy_unittest.cc]: this is not what you check. You check that the vector is sorted.
I am not too keen on this test. rather, you should check that *every* element in the training set is smaller than *every* element in the test set.
The manual test mentioned above would check that in a much simpler fashion, so I would just remove this one here

File: src/shogun/evaluation/TimeSeriesSplitting.cpp [hunk 30]
```diff
@@ -0,0 +1,107 @@
...
+void CTimeSeriesSplitting::set_min_subset_size(index_t min_size)
+{
+	index_t num_subsets = m_subset_indices->get_num_elements();
+	index_t num_labels = m_labels->get_num_labels();
+
+	/* min_size should be less than difference between number of labels
+	 * and split index of second last split */
+	REQUIRE(min_size > 0, "Minimum subset size should be atleast 1.")
+	REQUIRE(
+	    min_size < num_labels - (num_subsets - 1) * (num_labels / num_subsets),
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.cpp]: why not put this into a variable to make things cleaner?

File: src/shogun/evaluation/TimeSeriesSplitting.cpp [hunk 31]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+void CTimeSeriesSplitting::set_min_subset_size(index_t min_size)
+{
+	index_t num_subsets = m_subset_indices->get_num_elements();
+	index_t num_labels = m_labels->get_num_labels();
+
+	/* min_size should be less than difference between number of labels
+	 * and split index of second last split */
+	REQUIRE(min_size > 0, "Minimum subset size should be atleast 1.")
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.cpp]: missing newline "\n" at end of message

File: src/shogun/evaluation/TimeSeriesSplitting.cpp [hunk 32]
```diff
@@ -0,0 +1,107 @@
...
+
+	for (auto i = 0; i < num_subsets; ++i)
+	{
+		CDynamicArray<index_t>* current =
+		    (CDynamicArray<index_t>*)m_subset_indices->get_element(i);
+
+		if (i == num_subsets - 1)
+			split_index = indices.vlen - m_min_subset_size;
+		else
+			split_index = (i + 1) * (indices.vlen / num_subsets);
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.cpp]: are you sure this never suffers from integer division problems?

File: src/shogun/evaluation/TimeSeriesSplitting.h [hunk 33]
```diff
@@ -0,0 +1,98 @@
...
+#include <shogun/lib/config.h>
+
+namespace shogun
+{
+	class CLabels;
+
+	/** @brief Implements a timeseries splitting strategy for cross-validation.
+	 * The strategy builds a given number of subsets each filled with labels
+	 * indices greater than a split index. The split indices are \f$ c[N/K] \f$
+	 * where \f$N\f$ is number of labels,\f$K\f$ is number of subsets and
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.h]: Can you make this docs a bit more intuitive.

See https://github.com/scikit-learn/scikit-learn/blob/ea6ea815e1071bfb9de18953fea98b136a7fa8ed/sklearn/model_selection/_split.py#L676 for a very good example

File: src/shogun/evaluation/TimeSeriesSplitting.cpp [hunk 34]
```diff
@@ -0,0 +1,107 @@
...
+		/* filling current with indices on right end  */
+		for (auto k = split_index; k < indices.vlen; ++k)
+		{
+			current->append_element(indices.vector[k]);
+		}
+
+		SG_UNREF(current);
+	}
+
+	m_subset_indices->shuffle(m_rng);
```
Comment (karlnapf) [inline src/shogun/evaluation/TimeSeriesSplitting.cpp]: why this? No need

**Comment by Sahil333** [comment]
I will make these improvements in the next PR.

**Comment by karlnapf** [comment]
Great!
See also the recently opened issue on this #4049 . Maybe we can give this whole part of Shogun a facelift. If you are up for it :)

File: tests/unit/evaluation/SplittingStrategy_unittest.cc (hunk 24 above)
Comment (Sahil333) [reply tests/unit/evaluation/SplittingStrategy_unittest.cc]: I saw the FisherLDA_unittest.cc, great way. I will do like that.

File: tests/unit/evaluation/SplittingStrategy_unittest.cc (hunk 29 above)
Comment (Sahil333) [reply tests/unit/evaluation/SplittingStrategy_unittest.cc]: Oh, I see. Instead of checking that every element in the training set is smaller than every element in the test set, we can check min in the test set is greater than max in the training set. Will it do the trick?

Comment (karlnapf) [reply tests/unit/evaluation/SplittingStrategy_unittest.cc]: That will absolutely do it.


BUT, as I said, I think the best thing would be to write down an explicit case example that you can solve on paper...

File: src/shogun/evaluation/TimeSeriesSplitting.cpp (hunk 32 above)
Comment (Sahil333) [reply src/shogun/evaluation/TimeSeriesSplitting.cpp]: It shouldn't, as I need the floor of (N/K) and that's what the division would return.

File: src/shogun/evaluation/TimeSeriesSplitting.h (hunk 21 above)
Comment (Sahil333) [reply src/shogun/evaluation/TimeSeriesSplitting.h]: I need labels to know the number of indices needed to split.

File: tests/unit/evaluation/SplittingStrategy_unittest.cc (hunk 29 above)
Comment (Sahil333) [reply tests/unit/evaluation/SplittingStrategy_unittest.cc]: Shall I remove both the test with a single manual test using the fixtures?
